clients/drcachesim/tools/filter/record_filter.cpp

```diff
@@ -128,23 +137,99 @@ record_filter_t::parallel_shard_error(void *shard_data)
     return per_shard->error;
 }
 
+bool
+record_filter_t::write_trace_entry(per_shard_t *shard, const trace_entry_t &entry)
+{
+    if (!shard->writer->write((char *)&entry, sizeof(entry))) {
+        shard->error = "Failed to write to output file " + shard->output_path;
+        success_ = false;
+        return false;
+    }
+    ++shard->output_entry_count;
+    return true;
+}
+
 bool
 record_filter_t::parallel_shard_memref(void *shard_data, const trace_entry_t &entry)
 {
     per_shard_t *per_shard = reinterpret_cast<per_shard_t *>(shard_data);
+    ++per_shard->input_entry_count;
     bool output = true;
+    bool all_stop = true;
     for (int i = 0; i < (int)filters_.size(); ++i) {
-        if (!filters_[i]->parallel_shard_filter(entry, filter_shard_data_[i]))
+        bool stop = false;
+        if (!filters_[i]->parallel_shard_filter(entry, per_shard->filter_shard_data[i],
+                                                &stop))
             output = false;
+        if (stop)
+            per_shard->filter_stop[i] = true;
+        all_stop = all_stop && per_shard->filter_stop[i];
+    }
+
+    // Everything before the timestamp marker in the first unit header is part of the
+    // trace shard's header.
+    if (entry.type == TRACE_TYPE_MARKER && entry.size == TRACE_MARKER_TYPE_TIMESTAMP) {
+        per_shard->in_shard_header = false;
+    }
+
+    // We output everything in the trace shard's header.
+    if (per_shard->in_shard_header) {
+        output = true;
+    }
+
+    // Keep track of the last omitted unit header so that we can output it if and when
+    // we output a trace_entry_t from this unit.
```

Comment:
I am confused as to why we're doing all this work to support splitting up a trace: I thought that was a use case we did not consider high priority and would not spend time supporting in the near future.  I thought we were only going to support tools removing entries which is all we need for our near-term use cases.